js/ui/status/missionGame.js

```diff
@@ -0,0 +1,528 @@
+         * has finished scrolling */
+        const lastLabel = this._chatboxLabels.length === 0 ? null : this._chatboxLabels[this._chatboxLabels.length - 1];
+        const immediate = !lastLabel || lastLabel.complete;
+
+        /* Immediately put the label in this._chatboxLabels, however
+         * this does not mean it will be immediately added to the
+         * area. */
+        this._chatboxLabels.push(label);
+
+        if (immediate) {
+            _addLabelToChatboxArea(label, this._chatboxResultsArea);
+        } else {
+            lastLabel.connect('finished-scrolled', Lang.bind(this, function() {
+               _addLabelToChatboxArea(label, this._chatboxResultsArea);
+            }));
+        }
+    }
+});
+
+const MissionToolbox = new Lang.Class({
+    Name: 'MissionToolbox',
+    Extends: St.BoxLayout,
+    _init: function(params, parentMenu, service) {
+        this._service = service;
+
+        let parentParams = copyObject(params);
+        parentParams.vertical = true;
+        parentParams.name = 'missionToolbox';
+
+        this.parent(parentParams);
+
+        /* Game mode switch */
+        const missionSwitch = new PopupMenu.PopupSwitchMenuItem(_("Mission"), false);
+
+        /* Adventures and spells toolbox */
+        this._adventures = createSubMenuMenuItemWithFauxParent(_("Adventures"), parentMenu);
+        this._spells = createSubMenuMenuItemWithFauxParent(_("Spells"), parentMenu);
+        this._inventory = createSubMenuMenuItemWithFauxParent(_("Inventory"), parentMenu);
+
+        /* Add switches and toolbox items to hbox */
+        this.add(missionSwitch.actor);
+        addSubMenuItemToBox(this._adventures, this, parentMenu);
+        addSubMenuItemToBox(this._spells, this, parentMenu);
+        addSubMenuItemToBox(this._inventory, this, parentMenu);
+
+        /* When we get some new adventures or spells, add the menu items to the list again */
+        this._service.connect("discover-new-adventures", Lang.bind(this, function(chat, adventures) {
+            this._adventures.menu.removeAll();
+            adventures.forEach(Lang.bind(this, function(adventure) {
+                this._adventures.menu.addAction(adventure.desc, launchLessonAction(adventure.name));
+            }));
+        }));
+        this._service.connect("discover-new-spells", Lang.bind(this, function(chat, spells) {
+            this._spells.menu.removeAll();
+            spells.forEach(Lang.bind(this, function(spell) {
+                this._spells.menu.addAction(spell.desc, launchLessonAction(spell.name));
+            }));
+        }));
+        this._service.connect("discover-new-inventory-items", Lang.bind(this, function(chat, items) {
+            this._inventory.menu.removeAll();
+            items.forEach(Lang.bind(this, function(item) {
+                this._inventory.menu.addAction(item.name, function() {
+                });
+            }));
+        }));
+    }
+});
+
+function createSeparator() {
```

Comment:
It could be. I kept it here for the sake of not polluting the other parts of the shell codebase but otherwise I'd be in favour of abstracting it out.